internal/core/metadata/operators/device/profile/update.go

```diff
@@ -0,0 +1,141 @@
+// associated with it
+type UpdateDeviceProfileExecutor interface {
+	Execute() (contract.DeviceProfile, error)
+}
+
+// updateDeviceProfile updates device profiles by way of the operator pattern.
+type updateDeviceProfile struct {
+	database DeviceProfileUpdater
+	dp       contract.DeviceProfile
+}
+
+// Execute updates and persists the device profile.
+func (n updateDeviceProfile) Execute() (contract.DeviceProfile, error) {
+	// Check if the Device Profile exists
+	var to contract.DeviceProfile
+	// First try with ID
+	to, err := n.database.GetDeviceProfileById(n.dp.Id)
+	if err != nil {
+		// Try with name
+		to, err = n.database.GetDeviceProfileByName(n.dp.Name)
+		if err != nil {
+			return contract.DeviceProfile{}, errors2.NewErrDeviceProfileNotFound(n.dp.Id, n.dp.Name)
+		}
+	}
+
+	d, err := n.database.GetDevicesByProfileId(to.Id)
+	if err != nil {
+		return contract.DeviceProfile{}, err
+	}
+
+	if len(d) > 0 {
+		// LoggingClient.Debug(strconv.Itoa(len(d))+"devices are associated with the device profile named: " + to.Name)
+		// http.Error(w, "Conflict", http.StatusConflict)
+		return contract.DeviceProfile{}, errors2.NewErrDeviceProfileInvalidState(n.dp.Id, n.dp.Name, strconv.Itoa(len(d))+" devices are associated with the device profile")
+	}
+
+	p, err := n.database.GetProvisionWatchersByProfileId(to.Id)
+	if err != nil {
+		return contract.DeviceProfile{}, err
+	}
+
+	if len(p) > 0 {
+
+		return contract.DeviceProfile{}, errors2.NewErrDeviceProfileInvalidState(n.dp.Id, n.dp.Name, strconv.Itoa(len(p))+" provision watchers are associated with the device profile")
+	}
+
+	// Update the device profile fields based on the passed JSON
+	if err := n.updateDeviceProfileFields(n.dp, &to); err != nil {
+		return contract.DeviceProfile{}, err
+	}
+
+	if err := n.database.UpdateDeviceProfile(to); err != nil {
+		return contract.DeviceProfile{}, err
+	}
+
+	return to, nil
+}
+
+// NewUpdateDeviceProfileExecutor creates an UpdateDeviceProfileExecutor.
+func NewUpdateDeviceProfileExecutor(db DeviceProfileUpdater, dp contract.DeviceProfile) UpdateDeviceProfileExecutor {
+	return updateDeviceProfile{
+		database: db,
+		dp:       dp}
+}
+
+// Update the fields of the device profile
+// to - the device profile that was already in Mongo (whose fields we're updating)
+// from - the device profile that was passed in with the request
+func (n updateDeviceProfile) updateDeviceProfileFields(from contract.DeviceProfile, to *contract.DeviceProfile) error {
+	if from.Description != "" {
+		to.Description = from.Description
+	}
+
+	if from.Labels != nil {
+		to.Labels = from.Labels
+	}
+
+	if from.Manufacturer != "" {
+		to.Manufacturer = from.Manufacturer
+	}
+
+	if from.Model != "" {
+		to.Model = from.Model
+	}
+
+	if from.Origin != 0 {
+		to.Origin = from.Origin
+	}
+
+	if from.Name != "" {
+		to.Name = from.Name
+		// Names must be unique for each device profile
+		if err := n.checkDuplicateProfileNames(*to); err != nil {
+			return err
+		}
```

Comment:
In my opinion, this check should be moved into Execute() and performed before this function is even called. Once we get to invoking this function, all checks should be complete and the function does nothing but assign values to the new instance.

_Something to think about for later -- the creation of a new instance from old could actually be a function of the DeviceProfile model itself..._